packages/dom-adapters/src/caret/utils/useSelectionChange.ts

```diff
@@ -0,0 +1,143 @@
+import type { CaretAdapter } from '../CaretAdapter.js';
+import { createSingleton } from './singleton.js';
+
+/**
+ * Singleton that watches for document "selection change" event and delegates the provided callbacks to subscribers.
+ */
+export interface Subscriber {
+  /**
+   * Callback that will be called on "selection change" event.
+   *
+   * @param selection - current document selection
+   */
+  callback: (selection: Selection | null) => void;
+
+  /**
+   * Used to save context of the callback.
+   */
+  context: CaretAdapter;
+}
+
+/**
+ * Node that contains a selection.
+ *
+ * Now supports only contenteditable elements.
+ *
+ * @todo add support for native inputs
+ */
+export type InputWithCaret = HTMLElement;
+
+
+/**
+ * Utility composable that watches for document "selection change" event and delegates the provided callbacks to subscribers.
+ */
+export const useSelectionChange = createSingleton(() => {
+  /**
+   * User to iterate over all inputs and check if selection is related to them.
+   */
+  const inputsWatched: InputWithCaret[] = [];
+
+  /**
+   * WeakMap that stores subscribers for each input.
+   */
+  const subscribers = new WeakMap<InputWithCaret, Subscriber>();
+
+  /**
+   * Checks if selection is related to input
+   *
+   * @param selection - changed document selection
+   * @param input - input to check
+   */
+  function isSelectionRelatedToInput(selection: Selection | null, input: InputWithCaret): boolean {
+    if (!selection) {
+      return false;
+    }
+
+    const range = selection.getRangeAt(0);
+
+    return input.contains(range.startContainer) ?? false;
```

Comment:
Could you just add todo to think of cross-block selection later. To know where to look once we'll start working on it

Reply:
You can actually use `range.intersectsNode(input)`